internal/main.go: support for overriding metadata

AlexNPavel opened the pull request:
This adds the flag `set` which will set the specified attributes or
override the one's given by a provider.

Commit:
internal/main.go: support for overriding metadata

This adds the flag `set` which will set the specified attributes or
override the one's given by a provider.

bgilbert commented on internal/main.go:
```diff
@@ -66,6 +83,7 @@ func main() {
 	flag.StringVar(&flags.provider, "provider", "", "The name of the cloud provider")
 	flag.StringVar(&flags.sshKeys, "ssh-keys", "", "Update SSH keys for the given user")
 	flag.BoolVar(&flags.version, "version", false, "Print the version and exit")
+	flag.Var(&flags.overrides, "set", "Set an environment variable, overriding the default/detected variable")
```
- Should be in alphabetical order.
- The `attributes` help text calls them `metadata attributes`, so we should be consistent here.
- The help text should make clear that a key=value pair is wanted and that the option can be specified multiple times.

bgilbert commented on internal/main.go:
```diff
@@ -48,6 +48,22 @@ const (
+func (sm *StringMapFlag) Set(v string) error {
+	vals := strings.Split(v, "=")
+	if len(vals) != 2 {
+		fmt.Println("Could not split override set correctly")
```
This is potentially a user-visible error message and should clearly describe the problem.

bgilbert commented on internal/main.go:
```diff
@@ -48,6 +48,22 @@ const (
+	vals := strings.Split(v, "=")
+	if len(vals) != 2 {
+		fmt.Println("Could not split override set correctly")
+		return nil
```
Don't we want to fail in this case?

bgilbert commented on internal/main.go:
```diff
@@ -48,6 +48,22 @@ const (
+}
+
+func (sm *StringMapFlag) Set(v string) error {
+	vals := strings.Split(v, "=")
```
Values should be allowed to contain `=`, so this should be a `SplitN`.

bgilbert reviewed (changes requested):
@crawford @dgonyeo, thoughts on whether this is a good idea?  The motivator is to allow Vagrant to generate its own attributes (e.g. private IP address) that will be merged into whatever coreos-metadata produces on its own.

Commit:
review: improve Set function for StringMapFlag

This changes the function to return an error if a poorly formatted set
input is provided, give a more helpful error message, and use SplitN to
allow for values with `=` symbols in them

Commit:
review: organize flags alphabetically

Commit:
review: improve description for set flag

bgilbert commented on internal/main.go:
```diff
@@ -57,13 +73,15 @@ func main() {
 	flag.StringVar(&flags.hostname, "hostname", "", "The file into which the hostname should be written")
 	flag.StringVar(&flags.networkUnits, "network-units", "", "The directory into which network units are written")
 	flag.StringVar(&flags.provider, "provider", "", "The name of the cloud provider")
+	flag.Var(&flags.overrides, "set", "Set a metadata attribute, potentially overriding the attribute given by a provider. This flag can be specified multiple times and expects a key=value pair. Eg: VIRTUALBOX_IPV4_PRIVATE=172.17.4.101")
```
tiny nit: should be `e.g.`

bgilbert commented on internal/main.go:
```diff
@@ -48,6 +48,22 @@ const (
+	vals := strings.SplitN(v, "=", 1)
+	if len(vals) != 2 {
+		fmt.Printf("Incorrectly formatted -set argument: %s\n", v)
+		return fmt.Errorf("Incorrectly formatted -set argument: %s", v)
```
I get:

```
$ bin/coreos-metadata -set foo
Incorrectly formatted -set argument: foo
invalid value "foo" for flag -set: Incorrectly formatted -set argument: foo
Usage of bin/coreos-metadata:
[...]
```

Commit:
review: remove duplicate error printing for -set

This also fixes a tiny typo in the description for -set
